biohansel/create/cluster_generator.py

```diff
@@ -0,0 +1,236 @@
+                        that subtype
+    """
+
+    output_subtypes = {threshold: {} for threshold in clusters_genomes_dict.keys()}
+    sorted_thresholds = sorted(clusters_genomes_dict.keys())
+    # initialize top level subtypes
+    output_subtypes[sorted_thresholds[0]] = {str(cluster): genomes for cluster, genomes in
+                                             clusters_genomes_dict[sorted_thresholds[0]].items()}
+
+    for threshold_index in range(1, len(sorted_thresholds)):
+        parent_hc_clusters = output_subtypes[sorted_thresholds[threshold_index - 1]]
+
+        threshold = sorted_thresholds[threshold_index]
+        cluster_genomes = clusters_genomes_dict[threshold]
+
+        for subtype, parent_genomes in parent_hc_clusters.items():
+            if len(parent_genomes) < min_group_size:
+                continue
+            subclade = 1
+            for _, child_genomes in cluster_genomes.items():
+                if len(child_genomes) < min_group_size:
+                    continue
+
+                if parent_genomes == child_genomes:
+                    output_subtypes[threshold][subtype] = parent_genomes
+                elif child_genomes.issubset(parent_genomes):
+                    new_subgroup_name = f'{subtype}.{subclade}'
+                    if len(parent_genomes - child_genomes) >= min_group_size:
+                        output_subtypes[threshold][new_subgroup_name] = child_genomes
+                    else:
+                        output_subtypes[threshold][subtype] = child_genomes
+                    subclade += 1
+
+    return output_subtypes
+
+
+def expand_sets(cluster_dict: Dict[str, Dict[str, Set[str]]]) -> Dict[str, Dict[str, str]]:
+    """
+    Fills in the rest of the cells in the dataframe that had previously been unfilled
+
+    Args:
+        cluster_dict: a dictionary that contains the cluster groupings of each genome with each value being sets of
+                    genomes
+
+    Returns:
+        modified_cluster_dict: A dictionary with values beings filled for all cells
+
+    """
+
+    modified_cluster_dict = {}
+    for threshold, groupings in cluster_dict.items():
+        threshold_dict = {}
+        for grouping, genomes in groupings.items():
+            for genome in genomes:
+                threshold_dict[genome] = grouping
+        modified_cluster_dict[threshold] = threshold_dict
+
+    return modified_cluster_dict
+
+
+def row_subtype(curr_genome: pd.Series) -> str:
+    """
+    Provides the last valid subtype for that particular genome, as that's the last valid branch point
+
+    Args:
+        curr_genome: the current subtypes for that particular genome
+
+    Returns:
+        row_unique[-1]: the last valid subtype name for that genome
+
+    """
+
+    row_unique = curr_genome.unique()
+    row_unique = row_unique[(row_unique != '') & (~pd.isnull(row_unique))]
+
+    return row_unique[-1]
```

Comment:
Are you sure the last unique item will be the final subtype for each row? It may be a good idea to sort by string length to be safe.

Reply:
addressed in 99966c049e443fa7a170738ee145b3be2d7f64d4